### PR TITLE
Step 2: Add Namespace to Airbyte Protocol

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
*Describe what the change is solving*
Same as  #1993 to implement  #1921

## How
*Describe the solution*

This introduces a new object `AirbyteStreamName` in both the API and the Airbyte Protocol to represent namespace and table name in a single struct.
The old `name` of `AirbyteStream` is kept as it was but the new fields are properly populated with separate namespace/table names.

So for example, we would have catalog produced by the JDBC sources:
- Stream name: public.users (like before)
- AirbyteStreamName.namespace: public (new)
- AirbyteStreamName.name: users (new)

Non-jdbc sources are now producing:
- Stream name: users (like before)
- AirbyteStreamName.namespace:  (empty)
- AirbyteStreamName.name: users (same as stream name)

These new fields won't be used right away so they shouldn't introduce any change to the product until the last step is implemented. 

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*
- [ ] *Verify Acceptance tests* (depends on new docker images)

## Recommended reading order
1. `airbyte-api/src/main/openapi/config.yaml`
1. `airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml`
2. `airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java`
3. `airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java`
1. the rest

### CODE REVIEW

**Comment by ChristopheDuong** [comment]
When we make changes to the Airbyte Protocol, old source connectors are incompatible with newer versions of the protocol:

If I run a connector with a version from before my change to the protocol where I added the `stream_name` field, the validation of the Catalog from the old connector code will fail:

```
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "stream_name" (class io.airbyte.protocol.models.AirbyteStream), not marked as ignorable (5 known properties: "json_schema", "name", "supported_sync_modes", "default_cursor_field", "source_defined_cursor"]) 
```

[logs-7-0.txt](https://github.com/airbytehq/airbyte/files/6051255/logs-7-0.txt)

File: airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationTest.java
Comment (michel-tricot) [inline airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationTest.java:113]: could we create a helper: `CatalogHelpers.streamName(string name)` or `CatalogHelpers.streamName(string namespace, string name)` that returns an `AirbyteStreamName` and pass that object to `createConfiguredAirbyteStream` instead? Code will look better and it is just a matter of changing the constants type now.

File: airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/CatalogHelpers.java
Comment (michel-tricot) [inline airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/CatalogHelpers.java:44]: See my comment above. We should always manipulate namespace/streamname as AirbyteStreamName, never as two separate entities

File: airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java:37]: streamname should be the first parameter since it is the one that will remain. 

Can you add a Preconditions that checks that only one of the two is set? (assuming this is correct)

File: airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java:49]: same here.

I would also add a TODO that using string is deprecated

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java
Comment (michel-tricot) [inline airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:285]: should we unset this?

File: airbyte-workers/src/test/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteSourceTest.java
Comment (michel-tricot) [inline airbyte-workers/src/test/java/io/airbyte/workers/protocols/airbyte/DefaultAirbyteSourceTest.java:80]: Use the catalog helper  here

**Review by michel-tricot — COMMENTED** [review]
Looks really cool!

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java
Comment (ChristopheDuong) [reply airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:285]: No, this is required as we are currently backward compatible and everything keeps working as it was before.

The new AirbyteStreamName field is not actually influencing the stream names for the moment

File: airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java
Comment (ChristopheDuong) [reply airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java]: for the moment both are set, sometimes the StreamName object could be null (if it was serialized before introducing the namespace field)

File: airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlSourceTest.java
Comment (michel-tricot) [inline airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlSourceTest.java:53]: shouldn't the stream name just be tablename?

File: airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java:127]: NIT You're missing `{`

File: airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/converters/CatalogConverter.java:126]: this variable name is super confusing. I think you should just call it `name`

**Review by michel-tricot — APPROVED** [review]
Looks good! can you also get a review from someone else in the team before you merge?

File: airbyte-api/src/main/openapi/config.yaml
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:1741]:
```suggestion
          description: Stream's name (deprecated, will be replaced by streamName).
```

File: airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml
Comment (cgardens) [inline airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:131]: would it be simpler if instead of adding this `AirbyteStreamName` object, you just added namespace as a property to this top level object? then we don't need to deprecate `name`? it also saves us having to ever have name and stream_name at the same time.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:138]: can you add your github user name, or something else identifying when you add todos?

File: airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml
Comment (cgardens) [reply airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:131]: i got to think it makes the backwards compatibility story way simpler too.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:446]: does this all work okay with MySQL (which doesn't have schemas)?

**Review by cgardens — APPROVED** [review]
this seems like it should work okay (as long as all of the tests for the jdbc sources are passing).

I am not convinced that adding the `AirbyteStreamName` adds anything over just adding a `namespace` field. It seems like the extra object is kinda a headache. I think it would also be easier from a backwards compatibility point of view.

**Comment by cgardens** [comment]
what i'm describing won't work, because you're changing behavior in the sources before the destinations are compatible with namespaces so you're a little boxed. if you were just doing destinations first, i think you could the in place replacement (i.e. just add namespace). in that process each destination would be made to support namespaces first when they were present. each one could be migrated. once that was done, we could start migrating sources, and as they had non null values for namespaces the destinations would respect them). i don't think i love having this extra object, but it would have to be a pretty execution plan to get around it. 

probably fine to keep it as is, but wanted to at least offer this different perspective.

File: airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml
Comment (ChristopheDuong) [reply airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:131]: in my last PR,  @michel-tricot was mentioning this: https://github.com/airbytehq/airbyte/pull/1993#discussion_r572588441

Did I misunderstand his comment?

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:446]: Yes, I think we already figured it out when we made the PR to discover tables from another schema than just the `public` one.

File: airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlSourceTest.java
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlSourceTest.java:53]: The previous STREAM_NAME was "dbo.id_and_name"

We didn't introduce big changes on how things work so the stream name is still the same as before (but now it is also just "parsed" and split in two in the new AirbyteStreamName struct)

File: airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml
Comment (cgardens) [reply airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:131]: kk. i don't think you misunderstood.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:446]: kk

**Pull request closed without merging**
